pkg/apis/serving/v1alpha1/domainmapping_validation.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+)
+
+// Validate makes sure that DomainMapping is properly configured.
+func (dm *DomainMapping) Validate(_ context.Context) *apis.FieldError {
+	return validateMetadata(dm.ObjectMeta).ViaField("metadata").
+		Also(validateSpec(dm.Spec, dm.ObjectMeta).ViaField("spec"))
+}
+
+// validateSpec validates the Spec of a DomainMapping.
+func validateSpec(spec DomainMappingSpec, md metav1.ObjectMeta) (errs *apis.FieldError) {
+	return errs.Also(validateRef(spec.Ref, md).ViaField("ref"))
+}
+
+// validateRef validates the Ref field of a DomainMapping.
+func validateRef(ref duckv1.KReference, md metav1.ObjectMeta) (errs *apis.FieldError) {
+	if ref.Name == "" {
+		errs = errs.Also(apis.ErrMissingField("name"))
+	}
+
+	if ref.Namespace != "" && ref.Namespace != md.Namespace {
```

Comment:
Can `ref.Namespace` _be_ empty? Given you default it?

Reply:
mmmm dont know, maybe not :) tbh I just wanted to put some validation here to test stuff out, will swap this out for https://github.com/knative/pkg/blob/master/apis/duck/v1/knative_reference.go#L84 in next PR anyway when I fill in all the actual validation/defaulting :).

Reply:
no Conversion webhook because v1alpha1 is the only supported version, so I _think_ it doesn't make sense yet to allow any other version?